### ISSUE TITLE
ridden cart stopping occasionally, especially when going uphill

### ISSUE THREAD

**Issue by Konfuz** [joe7575/minecart#6]
I noticed that on long stretches of rail the cart might stopp and abort recording (which is annoying since I am trying to record a 6km long track). This seems to happen much more frequent when going uphill (no stairs below the track). 

### How to reproduce: 

Might be testable by building a circular track with a long upward slope, sitting in the cart and going for lunch while it makes its rounds. 

### Thesis:

I suspect that the engine reports the cart velocity and pitch as zero on occasion (perhaps right after set_pos?), leading to the stop-routine being executed. This may be a result of collision detection which would explain the much higher frequency when going uphill. Perhaps a fix might be to wait a few steps and then check again whether or not the cart is moving. I'll try to look into it myself (time and drive permitting) but thought to let you know. 

### Additions (edits) 

- Only occurs on a real server with network latency
- Uphill I'd say about every 250 nodes on average. 
- Going downhill its rare, I noticed the cart stops 1 node below where it should be (see 
![screenshot_20201105_204824](https://user-images.githubusercontent.com/11818151/98289767-f593c100-1fa8-11eb-9e42-e8dc1750e5c4.png)

**Comment by joe7575** [comment]
Thanks for the report. I will look into the code the next days.

**Comment by Konfuz** [comment]
I was not able to reproduce it in singleplayer so far, this adds to my suspicion that it is a network-related race-condition

**Comment by joe7575** [comment]
I can't reproduce it, too.
But it happened, when a player (you) was attached, or does it happen also without a "driver"?

**Comment by Konfuz** [comment]
So far we (multiple players) have only observed it with players attached, it rarely happens - but it happens about every 250 nodes when going uphill, which becomes an issue when you are trying to record several kilometers of track. In order to check without players I'd need to run along with the cart for a long distance :)

**Comment by joe7575** [comment]
When it happens, does it always happen at the same position?
We had this in the past, the cart always stopped at the same uphill position. We replaced the rails at that position and the bug was gone.

**Comment by Konfuz** [comment]
I spent last night with another computer scientist trying to connect our 2.6km long track-ramp from bottom to top and made some really weird observations. First off: Yes the carts do stop even when no player is attached. I had a few "runaway" cars that did stop right before mine did in the same spot. 

Since there was no way to make the entire tour we started to do little trips from buffer to buffer ~250m apart was our plan. 

First  the cart did stop in the same place like 10 times about 200m from the start. Replacing rails, changing nodes around. Nothing helped - until I built in a curve in the spot. After that the cart was passing the spot (still some stoppages, but very rarely and random) I rebuilt the original route afterwards and the cart started stopping again. Nothing special about the place and I did not notice anything special about the x,y,z position either. 

This led to the observation that building curves did reduce the amount of stoppages significantly especially where they happend in roughly the same place over and over again. There was a 200m long route segment where I had to put a curve every 20m as the cart would reliably stop about 20 nodes after the last curve over and over again. But after the next buffer suddenly and for no apparent reason there where no stoppages at all and I could drive about 1.2 km without stoppage to the end of our route. 

We tried to find any kind of global pattern, but found nothing that held to closer inspection. I can only say two things for sure: 1. changing the direction every now and then seems to help. 2. The whole thing drives me nuts. In the end, after we constructed a lot of new curves I managed to record the entire "ramp" from bottom to top (though later rides proved that I still was extremely lucky).

**Comment by joe7575** [comment]
Thank you for your extensive report.

I was able to generate these stops on my server too. I just built a small ring on a slope and on one day there were no issues but yesterday the cart stopped several times. It seems that it is some kind of race condition depending on the server load or other influences.
I will try to reproduce it on my development PC

**Comment by joe7575** [comment]
I was not able to reproduce the stops on by development PC, even running a dedicated server and several clients.
But today the stops don't occur on my server either, I guess the server load is lower than yesterday.
Are you interested to add some debugging code to generate logging entries and send me the log file outputs when the cart stopped?

**Comment by Konfuz** [comment]
I am not the server admin myself and depend on a third party for iteration. I'll keep looking into it - but I have plenty of other stuff going on (Thank you for taking your time to maintain your mod btw! I know you don't have to) 

I remember that with the original cart mod we had similar issues as well - in particular on long tracks and with poor server performance (we had `sqlite_synchronous = 2` which led to terrible performance) however instead of stopping the cart just detached from the player and ran away. Also there was glitching into walls which still happens with your mod on occasion, but rarely enough to ignore it. I suspect issues with the original code rather than something you introduced yourself.

**Comment by joe7575** [comment]
I changed the code. So give it a try and let me know if the bug is fixed.

**Comment by Konfuz** [comment]
Whatever you did - I think you fixed it. Just rode many km of track uphill without problems (where I had issues every 250 nodes before). sometimes I got some rather serious rubberbanding, but I dont think I can blame you for that (at least not in good conscience). 

One time I had a rather odd occurrence where the cart changed direction - I assume when it hit the terminal bumper while rubberbanding, and way later I had an instant where the server crashed right as I took a turn with the cart (again while rubberbanding - I'll ask for the server log) But as far as I can tell - good job :+1:  I'll open a new ticket if I can reproduce any of the odd behaviour but it seems rare enough to ignore.
